### ISSUE TITLE
Recursive invariants

### ISSUE THREAD

**Issue by davidcok** [OpenJML/OpenJML#478]
I’ve managed to sanitize the files causing the bug(s). I am attaching them. ByteBuffer.jml is a JML specification for java.nio.ByteBuffer. We may contribute this to the OpenJML project once we clean this up.

If you try to verify SBB.java (contained in foo.zip) you will see the following errors:

foo/SBB.java:21: error: Recursive attempt to assert or assume invariants - use pure helper methods: foo.SBB

```
public static SBB allocateDirect(int n) {

               ^
```

error: An error while executing a proof script for allocateDirect: (error "line 1021 column 7730: invalid function application, sort mismatch on argument at position 2")

2 errors

I haven’t been able to figure out what the recursive attempt is due to. Any help would be appreciated.

Also, your critique of Buffer.jml and ByteBuffer.jml would be helpful.

Thanks!
Serdar
[foo.zip](https://github.com/OpenJML/OpenJML/files/501203/foo.zip)

**Comment by davidcok** [comment]
The second error is fixed in commit 90636cdf24b203812459b45fb7fbb7fc68d9f1b1 and will be in version 0.7.3.  The first error is not yet reproduced.

**Comment by davidcok** [comment]
Confirmed fixed in 0.8.19
